lib/pbench/test/unit/server/test_schema.py

```diff
@@ -126,10 +155,88 @@ def test_invalid_required(self, test):
         ),
     )
     def test_invalid_optional(self, test):
+        """
+        Test parameter validation of an optional parameter
+        """
         x = Parameter("data", ParamType.STRING, required=False)
         json, expected = test
         assert x.invalid(json) is expected
 
+    @pytest.mark.parametrize(
+        "test",
+        (
+            ({"data": "yes"}, "Yes"),
+            ({"data": "NO"}, "No"),
+            ({"data": "MaYbE"}, "MAYBE"),
+            ({"data": "Maybe"}, "MAYBE"),
+        ),
+    )
+    def test_keyword_normalization(self, test):
+        """
+        Test parameter normalization for a keyword parameter.
+        """
+        x = Parameter("data", ParamType.KEYWORD, keywords=["Yes", "No", "MAYBE"])
+        json, expected = test
+        assert x.normalize(json.get("data")) == expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            {"data": "I'm not sure"},
+            {"data": "yes!"},
+            {"data": "ebyam"},
+            {"data": ["Yes"]},
+            {"data": False},
+        ),
+    )
+    def test_invalid_keyword(self, test):
+        """
+        Test parameter normalization for invalid keyword parameter values.
+        """
+        x = Parameter("data", ParamType.KEYWORD, keywords=["Yes", "No"], required=False)
+        with pytest.raises(ConversionError) as exc:
+            x.normalize(test.get("data"))
+        assert str(exc).find(str(test.get("data"))) > 0
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, None, {"data": ["yes", "no"]}, ["yes", "no"]),
+            (ParamType.KEYWORD, ["Yes", "No"], {"data": ["YeS", "nO"]}, ["Yes", "No"]),
+            (
+                ParamType.ACCESS,
+                None,
+                {"data": ["Public", "PRIVATE"]},
+                ["public", "private"],
+            ),
+        ),
+    )
+    def test_list_normalization(self, test):
+        """
+        Test parameter normalization for a list parameter.
+        """
+        type, keys, json, expected = test
+        x = Parameter("data", ParamType.LIST, keywords=keys, element_type=type)
+        assert x.normalize(json.get("data")) == expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, None, {"data": [False, 1]}),
+            (ParamType.KEYWORD, ["Yes", "No"], {"data": ["maybe", "nO"]}),
+            (ParamType.ACCESS, None, {"data": ["publish", "PRIVATE"]}),
```

Comment:
I recommend something more obviously wrong than `"publish"`....  Yes, I know the point of this is to catch subtle mistakes, but it's not necessary for the test for the keyword to be subtly wrong.  If you want the first five charcters to match, that's fine...so change the last two to `x`'s or something which the reader won't gloss over.

Reply:
Can't go more wrong than Sauron, I suppose!  😁